### PR TITLE
i#5505 kernel trace: refactor the build system of PT-related libraries

### PR DESCRIPTION
**Pull request by dolanzhao**
Adds variables in cmake file to control the build process of drpttracer and libdrpt2ir separately. Dr's build system currently supports three modes:

(1) DR will not build drpttracer and libdrpt2ir on non-x86_64 architectures or non-Linux systems.
(2) DR will build drpttracer and libdrpt2ir on x86_64 Linux systems if the submodule libipt is initialized.
(3) DR will only build drpttracer on x86_64 Linux systems if libipt is not initialized.

Issue: #5505

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1217,17 +1217,26 @@ else ()
...
+set(BUILD_PT_POST_PROCESSOR OFF)
 set(proc_supports_pt OFF)
+# Right now we only build PT related libraries on Linux x86_64.
 if (LINUX AND X86 AND X64)
-  # Right now we only build libipt on Linux x86_64.
-  # We define a variable to indicate whether DR will build libipt.
-  # TODO i#5505: In the next PR, we need to check whether the libipt submodule is initialized
-  # and print a warning message when the libipt directory is empty.
-  set(HAVE_LIBIPT ON)
+  # We will always build the PT tracer no matter whether the system supports PT.
```
Comment (abhinav92003) [inline CMakeLists.txt:1225]: Add to the comment the rationale behind that: users may want to use the built binaries on a different system that does support PT.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1217,17 +1217,26 @@ else ()
...
   check_intel_pt_support(proc_supports_pt)
-else ()
-  set(HAVE_LIBIPT OFF)
-endif ()
+  # libdrpt2ir and drpt2trace depend on libipt. If the libipt submodule is not initialized,
+  # cmake will report an warning here.
+  if (EXISTS "${PROJECT_SOURCE_DIR}/third_party/libipt/.git" )
+    set(BUILD_PT_POST_PROCESSOR ON)
+  else ()
+    message(STATUS "libipt submodule is not initialized. libdrpt2ir and drpt2trace will not be built.")
```
Comment (abhinav92003) [inline CMakeLists.txt:1235]: When we integrate with raw2trace, will it be an issue if we don't build libdrpt2ir and drpt2trace?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1217,17 +1217,26 @@ else ()
...
+  endif ()
+else (LINUX AND X86 AND X64)
+  message(STATUS "PT related libraries only supported on Linux x86_64")
```
Comment (abhinav92003) [inline CMakeLists.txt:1238]: In this case, some DR dox will contain broken references right?

Comment (derekbruening) [reply CMakeLists.txt:1238]: Again, we can separate building the code from building the docs, which is what we already do for extension libraries and platform-specific functions.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1217,17 +1217,26 @@ else ()
   set(HAVE_HALF_FLOAT OFF)
```
Comment (derekbruening) [inline CMakeLists.txt:1217]: s/Dr /DR /g in the commit description

File: CMakeLists.txt [hunk 5]
```diff
@@ -1217,17 +1217,26 @@ else ()
...
+# Right now we only build PT related libraries on Linux x86_64.
 if (LINUX AND X86 AND X64)
-  # Right now we only build libipt on Linux x86_64.
-  # We define a variable to indicate whether DR will build libipt.
-  # TODO i#5505: In the next PR, we need to check whether the libipt submodule is initialized
-  # and print a warning message when the libipt directory is empty.
-  set(HAVE_LIBIPT ON)
+  # We will always build the PT tracer no matter whether the system supports PT.
+  set(BUILD_PT_TRACER ON)
+  # Check if the building machine support Intel PT. This function only checks if PT-related tests
```
Comment (derekbruening) [inline CMakeLists.txt:1227]: grammar: s/support/supports/

File: CMakeLists.txt [hunk 6]
```diff
@@ -1217,17 +1217,26 @@ else ()
...
+  # need to be built. PT-capable binaries can be built on any system.
   check_intel_pt_support(proc_supports_pt)
-else ()
-  set(HAVE_LIBIPT OFF)
-endif ()
+  # libdrpt2ir and drpt2trace depend on libipt. If the libipt submodule is not initialized,
+  # cmake will report an warning here.
```
Comment (derekbruening) [inline CMakeLists.txt:1231]: grammar: s/an/a/

File: suite/tests/CMakeLists.txt [hunk 7]
```diff
@@ -4170,8 +4170,8 @@ endif (BUILD_CLIENTS)
 # ##########################################################################
 # drpt2trace tests
 # We use the libipt/script/perf-get-opts.bash script to get the options for drpt2trace's tests.
-if(BUILD_CLIENTS)
-  if(HAVE_LIBIPT)
+if (BUILD_CLIENTS)
+  if (BUILD_PT_POST_PROCESSOR)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:4174]: BUILD_PT_TRACER AND BUILD_PT_POST_PROCESSOR

**Review by derekbruening — APPROVED** [review]

File: suite/tests/CMakeLists.txt (hunk 7 above)
Comment (dolanzhao) [reply suite/tests/CMakeLists.txt:4174]: Done.

File: CMakeLists.txt (hunk 1 above)
Comment (dolanzhao) [reply CMakeLists.txt:1225]: Done.

File: CMakeLists.txt (hunk 6 above)
Comment (dolanzhao) [reply CMakeLists.txt:1231]: Done.

File: CMakeLists.txt (hunk 5 above)
Comment (dolanzhao) [reply CMakeLists.txt:1227]: Done.

File: CMakeLists.txt (hunk 4 above)
Comment (dolanzhao) [reply CMakeLists.txt:1217]: Done.

File: CMakeLists.txt (hunk 2 above)
Comment (dolanzhao) [reply CMakeLists.txt:1235]: Could we add a cmakedefine in configure.cmake.h?

File: CMakeLists.txt (hunk 3 above)
Comment (dolanzhao) [reply CMakeLists.txt:1238]: Will do it in next PR.

Comment (dolanzhao) [reply CMakeLists.txt:1238]: > In this case, some DR dox will contain broken references right?

This is a bug in my code's comments. I have fixed it. Sorry forgot to sync with you.

File: CMakeLists.txt (hunk 2 above)
Comment (abhinav92003) [reply CMakeLists.txt:1236]: Sure that works.

**Review by abhinav92003 — APPROVED** [review]

File: CMakeLists.txt [hunk 8]
```diff
@@ -1222,13 +1222,14 @@ set(BUILD_PT_POST_PROCESSOR OFF)
 set(proc_supports_pt OFF)
 # Right now we only build PT related libraries on Linux x86_64.
 if (LINUX AND X86 AND X64)
-  # We will always build the PT tracer no matter whether the system supports PT.
+  # We will always build the PT tracer on Linux X86_64 no matter whether the system supports PT.
```
Comment (derekbruening) [inline CMakeLists.txt:1225]: style violation: Line too long.

File: CMakeLists.txt [hunk 9]
```diff
@@ -1222,13 +1222,14 @@ set(BUILD_PT_POST_PROCESSOR OFF)
...
+  # Because users may want to use the built binaries on a different system that does support PT.
```
Comment (derekbruening) [inline CMakeLists.txt:1226]: Ditto.

File: CMakeLists.txt [hunk 10]
```diff
@@ -1222,13 +1222,14 @@ set(BUILD_PT_POST_PROCESSOR OFF)
...
   set(BUILD_PT_TRACER ON)
-  # Check if the building machine support Intel PT. This function only checks if PT-related tests
+  # Check if the building machine supports Intel PT. This function only checks if PT-related tests
```
Comment (derekbruening) [inline CMakeLists.txt:1228]: Ditto.

File: CMakeLists.txt (hunk 9 above)
Comment (derekbruening) [inline CMakeLists.txt:1226]: Do not make this a separate sentence: s/PT. Because/PT, because/

File: make/configure.cmake.h [hunk 11]
```diff
@@ -110,6 +110,8 @@
 #cmakedefine ANNOTATIONS
 #cmakedefine HAVE_RSEQ
 #cmakedefine HAVE_HALF_FLOAT
+#cmakedefine BUILD_PT_POST_PROCESSOR
```
Comment (derekbruening) [inline make/configure.cmake.h:113]: These are added but not used anywhere: I assume added as a mistake and should be removed.

**Review by derekbruening — APPROVED** [review]

File: make/configure.cmake.h (hunk 11 above)
Comment (abhinav92003) [reply make/configure.cmake.h:113]: These are intended to be used to conditionally compile code that invokes drpttracer and drpt2ir during online trace and offline post-processing. I suppose they can be added in the next PR when we actually add that new code.

File: CMakeLists.txt (hunk 8 above)
Comment (dolanzhao) [reply CMakeLists.txt:1225]: Done.

File: CMakeLists.txt (hunk 9 above)
Comment (dolanzhao) [reply CMakeLists.txt:1226]: Done.

File: CMakeLists.txt (hunk 10 above)
Comment (dolanzhao) [reply CMakeLists.txt:1228]: Done.

File: CMakeLists.txt (hunk 9 above)
Comment (dolanzhao) [reply CMakeLists.txt:1226]: Done.

File: make/configure.cmake.h (hunk 11 above)
Comment (dolanzhao) [reply make/configure.cmake.h:113]: I removed them. And I will add them in next PR.

**Pull request merged**